src/nbl/video/IPhysicalDevice.cpp

```diff
@@ -90,4 +90,375 @@ bool IPhysicalDevice::validateLogicalDeviceCreation(const ILogicalDevice::SCreat
+        uint32_t bitshft = 2;
+        return core::Float16Compressor::decompress(1 << bitshft) - 0.0;
+    }
+    default: return asset::getFormatPrecision(format, channel, 0.f);
+    }
+}
+
+// Returns true if a has higher precision & value range than b
+bool higherPrecisionOrValueRange(asset::E_FORMAT a, asset::E_FORMAT b, uint32_t srcChannels)
+{
+    auto c = 0;
+    while (c < srcChannels)
+    {
+        // break if a has less precision (higher precision value from getFormatPrecision) than b
+        if (getMinFormatPrecision(a, c) > getMinFormatPrecision(b, c))
+            break;
+        // break if a has less range than b
+        if (asset::getFormatMinValue<double>(a, c) > asset::getFormatMinValue<double>(b, c)
+            || asset::getFormatMaxValue<double>(a, c) < asset::getFormatMinValue<double>(b, c))
+            break;
+
+        c++;
+    }
+    // If we hit a break in the while loop, return false
+    return c == srcChannels;
+}
+
+// Rules for promotion:
+// - Cannot convert to block or planar format
+// - Aspects: Preserved or added
+// - Channel count: Preserved or increased
+// - Data range: Preserved or increased (per channel)
+// - Data precision: Preserved or improved (per channel)
+//     - Bit depth when comparing non srgb
+// If there are multiple matches: Pick smallest texel block
+// srcFormat can't be in validFormats (no promotion should be made if the format itself is valid)
+asset::E_FORMAT narrowDownFormatPromotion(const core::unordered_set<asset::E_FORMAT>& validFormats, asset::E_FORMAT srcFormat)
+{
+    asset::E_FORMAT smallestTexelBlock = asset::E_FORMAT::EF_UNKNOWN;
+    uint32_t smallestTexelBlockSize = -1;
+
+    auto srcChannels = asset::getFormatChannelCount(srcFormat);
+    assert(srcChannels);
+    auto srcAspects = getImageAspects(srcFormat);
+    bool srcIntFormat = asset::isIntegerFormat(srcFormat);
+
+    float srcPrecision[4];
+    double srcMinVal[4];
+    double srcMaxVal[4];
+    for (uint32_t channel = 0; channel < srcChannels; channel++)
+    {
+        srcPrecision[channel] = getMinFormatPrecision(srcFormat, channel);
+        srcMinVal[channel] = asset::getFormatMinValue<double>(srcFormat, channel);
+        srcMaxVal[channel] = asset::getFormatMaxValue<double>(srcFormat, channel);
+    }
+
+    // Better way to iterate the bitflags here?
+    for (auto iter = validFormats.begin(); iter != validFormats.end(); iter++)
+    {
+        asset::E_FORMAT f = *iter;
+
+        // Can't transcode to BC or planar
+        if (asset::isBlockCompressionFormat(f))
+            continue;
+        if (asset::isPlanarFormat(f))
+            continue;
+    
+        // Can't promote between int and normalized/float/scaled formats
+        if (srcIntFormat != asset::isIntegerFormat(f))
+            continue;
+
+        // Can't have less channels
+        auto dstChannels = asset::getFormatChannelCount(f);
+        if (dstChannels < srcChannels)
+            continue;
```

Comment:
could you hoist these checks and perform them before anything gets put into the `validFormats` set?